kubectl-pg/cmd/connect.go

```diff
@@ -0,0 +1,122 @@
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package cmd
+
+import (
+	"fmt"
+	"github.com/spf13/cobra"
+	"log"
+	"os"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/remotecommand"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	PostgresqlLister "github.com/zalando/postgres-operator/pkg/generated/clientset/versioned/typed/acid.zalan.do/v1"
+	"strconv"
+	"k8s.io/client-go/rest"
+)
+
+// connectCmd represents the kubectl pg connect command
+var connectCmd = &cobra.Command{
+	Use:   "connect",
+	Short: "A ",
+	Long: `A .`,
+	Run: func(cmd *cobra.Command, args []string) {
+		clusterName,_ := cmd.Flags().GetString("clusterName")
+		master,_ := cmd.Flags().GetBool("master")
+		replica,_ := cmd.Flags().GetString("replica")
+		psql,_ := cmd.Flags().GetBool("psql")
+		connect(clusterName,master,replica,psql)
+	},
+}
+
+func connect(clusterName string,master bool,replica string,psql bool) {
+	config := getConfig()
+	client,er := kubernetes.NewForConfig(config)
+	if er != nil {
+		log.Fatal(er)
+	}
+	postgresConfig, err := PostgresqlLister.NewForConfig(config)
+	postgresCluster, err := postgresConfig.Postgresqls(getCurrentNamespace()).Get(clusterName, metav1.GetOptions{})
+	if err != nil {
+		log.Fatal(err)
+	}
+	numOfInstances := postgresCluster.Spec.NumberOfInstances
+	var podName string
+	var podRole string
+	for ins:=0;ins < int(numOfInstances);ins++ {
+		pod,err := client.CoreV1().Pods(getCurrentNamespace()).Get(clusterName+"-"+strconv.Itoa(ins),metav1.GetOptions{})
+		if err != nil {
+			log.Fatal(err)
+		}
+		podRole = pod.Labels["spilo-role"]
+		if podRole == "master" && master {
+			podName = pod.Name
+			fmt.Printf("connected to %s with name %s\n",podRole, podName)
+			break
+		} else if podRole == "replica" &&  !master  && (pod.Name == replica || replica == "") {
+			podName = pod.Name
+			fmt.Printf("connected to %s with pod name as %s\n",podRole, podName)
+			break
+		}
+	}
+	execRequest := &rest.Request{}
+	if psql {
+		execRequest = client.CoreV1().RESTClient().Post().Resource("pods").
+			Name(podName).
+			Namespace("default").
+			SubResource("exec").
+			Param("container", "postgres").
+			Param("command", "psql").
+			Param("command", "-U").
+			Param("command", "postgres").
+			Param("stdin", "true").
+			Param("stdout", "true").
+			Param("stderr", "true").
+			Param("tty", "true")
+	} else {
+		execRequest = client.CoreV1().RESTClient().Post().Resource("pods").
+			Name(podName).
+			Namespace("default").
+			SubResource("exec").
+			Param("container", "postgres").
+			Param("command", "bash").
+			Param("stdin", "true").
+			Param("stdout", "true").
+			Param("stderr", "true").
+			Param("tty", "true")
+	}
```

Comment:
I feel this whole if - else block can be simplified. Only if `-p` is set you could append more params. Is this possible?

```
execRequest = client.CoreV1().RESTClient().Post().Resource("pods").
			Name(podName).
			...

if psql {
execRequest.Param("command", "psql")
  if user {
    execRequest.Param("command", "-U").Param("command", userName)
  }
}
```

Reply:
Yes we can do that if that's expected and what if an invalid username is specified

Reply:
The DB will give you an error then, no?

Reply:
There's another piece missing here: The database to connect to. When calling psql without any parameters it will usually try to connect to the `postgres` database. But, if you specify a user name, it will try to connect to a db with the same name which now leads to this error:
```
kubectl pg connect -c acid-minimal-cluster -m -p -u foo_user
connected to master with pod name as acid-minimal-cluster-0
psql: FATAL:  database "foo_user" does not exist
2019/08/14 17:08:55 command terminated with exit code 2
```

Reply:
When adding another parameter to specify the DB (`-d`) we could also drop the psql `-p` flag, I guess.

Reply:
Could reflect on my first comment about making this whole if else block a little nicer?